File: src/server/src/test/test_velvet_dawn/test_mechanics/test_selectors/test_filters.py
import velvet_dawn
from test.base_test import BaseTest
from velvet_dawn.dao import app
from velvet_dawn.dao.models.world_instance import WorldInstance
from velvet_dawn.mechanics import selectors


class TestFilteredSelectors(BaseTest):

    def test_filter_parsing(self):
        # Test selectors work
        selector = selectors.get_selector(
            "0", "unit[id=10,tag=dsa,range=100]>commanders[id=4,tag=10]")

        self.assertIsInstance(selector, selectors.SelectorUnit)
        self.assertIsInstance(selector.chained_selector, selectors.SelectorCommanders)

        self.assertIn('10', selector.filters.allowed_ids)
        self.assertIn('tag:dsa', selector.filters.allowed_tags)
        self.assertEqual(100, selector.filters.max_range)
        self.assertIsNone(selector.filters.min_range)

        self.assertIn('4', selector.chained_selector.filters.allowed_ids)
        self.assertIn('tag:10', selector.chained_selector.filters.allowed_tags)
        self.assertNotIn('10', selector.chained_selector.filters.allowed_tags)
        self.assertIsNone(selector.chained_selector.filters.max_range)
        self.assertIsNone(selector.chained_selector.filters.min_range)

    def test_filters_id(self):
        with app.app_context():
            self.prepare_game()

            selector_1 = selectors.get_selector('world', "units[tag=tag:test-tag1]")
            selector_2 = selectors.get_selector('world', "units[tag=x]")

            self.assertEqual(2, len(selector_1.get_selection(WorldInstance())))
            self.assertEqual(0, len(selector_2.get_selection(WorldInstance())))

    def test_filters_tags(self):
        with app.app_context():
            self.prepare_game()

            selector_1 = selectors.get_selector('world', "units[id=testing:commander]")
            selector_2 = selectors.get_selector('world', "units[id=testing:upgradable]")
            selector_3 = selectors.get_selector('world', "units[id=testing:commander, id=testing:upgradable]")

            self.assertEqual(2, len(selector_1.get_selection(WorldInstance())))
            self.assertEqual(1, len(selector_2.get_selection(WorldInstance())))
            self.assertEqual(3, len(selector_3.get_selection(WorldInstance())))

    def test_filters_max_range(self):
        with app.app_context():
            self.prepare_game()

            unit = velvet_dawn.units.get_unit_at_position(5, 0)

            selector_range_0 = selectors.get_selector(unit.entity_id, "tiles[range=0]")
            selector_range_1 = selectors.get_selector(unit.entity_id, "tiles[range=1]")
            selector_range_2 = selectors.get_selector(unit.entity_id, "tiles[range=2]")

            self.assertEqual(1, len(selector_range_0.get_selection(unit)))
            self.assertEqual(6, len(selector_range_1.get_selection(unit)))
            self.assertEqual(11, len(selector_range_2.get_selection(unit)))

    def test_filters_min_range(self):
        with app.app_context():
            self.prepare_game()

            unit = velvet_dawn.units.get_unit_at_position(5, 0)

            selector_range_0 = selectors.get_selector(unit.entity_id, "tiles[min-range=0]")
            selector_range_1 = selectors.get_selector(unit.entity_id, "tiles[min-range=1]")
            selector_range_2 = selectors.get_selector(unit.entity_id, "tiles[min-range=2]")

            self.assertEqual(77, len(selector_range_0.get_selection(unit)))
            self.assertEqual(76, len(selector_range_1.get_selection(unit)))
            self.assertEqual(71, len(selector_range_2.get_selection(unit)))

    def test_exclude_test_filter(self):
        with app.app_context():
            self.prepare_game()

            unit = velvet_dawn.units.get_unit_at_position(5, 0)

            selector_self = selectors.get_selector(unit.entity_id, "self[exclude-self]")
            selector_units = selectors.get_selector(unit.entity_id, "units[exclude-self]")

            self.assertEqual(0, len(selector_self.get_selection(unit)))
            self.assertEqual(3, len(selector_units.get_selection(unit)))  # only 3 of 4 units returned

File: src/server/test/velvetdawn/mechanics/selectors/TestFilters.java
package velvetdawn.mechanics.selectors;

import org.junit.Test;
import velvetdawn.BaseTest;
import velvetdawn.core.mechanics.selectors.SelectorUnit;
import velvetdawn.core.mechanics.selectors.Selectors;
import velvetdawn.core.models.Coordinate;
import velvetdawn.core.models.instances.WorldInstance;

import java.util.ArrayList;

import static junit.framework.TestCase.assertFalse;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotEquals;
import static org.junit.Assert.assertNull;
import static org.junit.Assert.assertTrue;


public class TestFilters extends BaseTest {

    @Test
    public void test_filter_parsing() throws Exception {
        var velvetDawn = this.prepareGame();

        // Test selectors work
        var selector = Selectors.get(
                velvetDawn,
                "0",
                "entity[id=10,tag=dsa,range=100]>entities[commander,id=4,tag=10]");

        assertTrue(selector instanceof SelectorUnit);

        assertTrue(selector.filters.allowedIds.contains("10"));
        assertTrue(selector.filters.allowedTags.contains("dsa"));
        assertFalse(selector.filters.commanderOnly);
        assertEquals(100, selector.filters.maxRange, 0);
        assertNull(selector.filters.minRange);

        assertTrue(selector.chainedSelector.filters.allowedIds.contains("4"));
        assertTrue(selector.chainedSelector.filters.allowedTags.contains("10"));
        assertTrue(selector.chainedSelector.filters.commanderOnly);
        assertNull(selector.chainedSelector.filters.maxRange);
        assertNull(selector.chainedSelector.filters.minRange);
    }

    @Test
    public void test_filters_id() throws Exception {
        var velvetDawn = this.prepareGame();

        var selector_1 = Selectors.get(velvetDawn, "world", "entities[tag=test-tag1]");
        var selector_2 = Selectors.get(velvetDawn, "world", "entities[tag=x]");

        assertEquals(2, selector_1.getChainedSelection(WorldInstance.getInstance()).size());
        assertEquals(0, selector_2.getChainedSelection(WorldInstance.getInstance()).size());
    }

    @Test
    public void test_filters_tags() throws Exception {
        var velvetDawn = this.prepareGame();

        var selector_1 = Selectors.get(velvetDawn, "world", "entities[id=testing:commander]");
        var selector_2 = Selectors.get(velvetDawn, "world", "entities[id=testing:upgradable]");
        var selector_3 = Selectors.get(velvetDawn, "world", "entities[id=testing:commander, id=testing:upgradable]");

        assertEquals(2, selector_1.getChainedSelection(WorldInstance.getInstance()).size());
        assertEquals(1, selector_2.getChainedSelection(WorldInstance.getInstance()).size());
        assertEquals(3, selector_3.getChainedSelection(WorldInstance.getInstance()).size());
    }

    @Test
    public void test_filters_max_range() throws Exception {
        var velvetDawn = this.prepareGame();

        var unit = velvetDawn.entities.getAtPosition(new Coordinate(5, 0)).get(0);

        var selector_range_0 = Selectors.get(velvetDawn, unit.instanceId, "tiles[range=0]");
        var selector_range_1 = Selectors.get(velvetDawn, unit.instanceId, "tiles[range=1]");
        var selector_range_2 = Selectors.get(velvetDawn, unit.instanceId, "tiles[range=2]");

        assertEquals(1, selector_range_0.getChainedSelection(unit).size());
        assertEquals(6, selector_range_1.getChainedSelection(unit).size());
        assertEquals(13, selector_range_2.getChainedSelection(unit).size());
    }

    @Test
    public void test_filters_min_range() throws Exception {
        var velvetDawn = this.prepareGame();

        var unit = velvetDawn.entities.getAtPosition(new Coordinate(5, 0)).get(0);

        var selector_range_0 = Selectors.get(velvetDawn, unit.datapackId, "tiles[min-range=0]");
        var selector_range_1 = Selectors.get(velvetDawn, unit.datapackId, "tiles[min-range=1]");
        var selector_range_2 = Selectors.get(velvetDawn, unit.datapackId, "tiles[min-range=2]");

        assertEquals(121, selector_range_0.getChainedSelection(unit).size());
        assertEquals(120, selector_range_1.getChainedSelection(unit).size());
        assertEquals(115, selector_range_2.getChainedSelection(unit).size());
    }

    @Test
    public void test_exclude_test_filter() throws Exception {
        var velvetDawn = this.prepareGame();

        var unit = velvetDawn.entities.getAtPosition(new Coordinate(5, 0)).get(0);

        var selector_self = Selectors.get(velvetDawn, unit.datapackId, "self[exclude-self]");
        var selector_units = Selectors.get(velvetDawn, unit.datapackId, "entities[exclude-self]");

        assertEquals(0, selector_self.getChainedSelection(unit).size());
        assertEquals(3, selector_units.getChainedSelection(unit).size());  // only 3 of 4 units returned
    }

    @Test
    public void test_selector_unit() throws Exception {
        var velvetDawn = this.prepareGame();

        var unit = velvetDawn.entities.getAtPosition(new Coordinate(5, 0)).get(0);

        var selectorClosestUnit = Selectors.get(velvetDawn, "0", "entities[closest]");
        var selectorClosestEnemy = Selectors.get(velvetDawn, "0", "enemies[closest]");
        var selectorClosestFriendly = Selectors.get(velvetDawn, "0", "friendlies[closest]");

        var closest_unit = selectorClosestUnit.getChainedSelection(unit);
        var closest_enemy = selectorClosestEnemy.getChainedSelection(unit);
        var closest_friendly = selectorClosestFriendly.getChainedSelection(unit);

        assertEquals(1, closest_unit.size());
        assertEquals(1, closest_enemy.size());
        assertEquals(1, closest_friendly.size());

        // Test that self is excluded
        assertNotEquals(new ArrayList<>(closest_unit).get(0).instanceId, unit.instanceId);
    }
}